src/irr/asset/CImageWriterOpenEXR.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+MIT License
+Copyright (c) 2019 AnastaZIuk
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "CImageWriterOpenEXR.h"
+
+#ifdef _IRR_COMPILE_WITH_OPENEXR_WRITER_
+
+#include "openexr/IlmBase/Imath/ImathBox.h"
+#include "openexr/OpenEXR/IlmImf/ImfOutputFile.h"
+#include "openexr/OpenEXR/IlmImf/ImfChannelList.h"
+#include "openexr/OpenEXR/IlmImf/ImfChannelListAttribute.h"
+#include "openexr/OpenEXR/IlmImf/ImfStringAttribute.h"
+#include "openexr/OpenEXR/IlmImf/ImfMatrixAttribute.h"
+#include "openexr/OpenEXR/IlmImf/ImfArray.h"
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+
+#include "openexr/OpenEXR/IlmImf/ImfNamespace.h"
+namespace IMF = OPENEXR_IMF_NAMESPACE;
+namespace IMATH = IMATH_NAMESPACE;
+
+namespace irr
+{
+	namespace asset
+	{
+		using namespace IMF;
+		using namespace IMATH;
+
+		constexpr uint8_t availableChannels = 4;
+
+		template<typename ilmType>
+		void createAndWriteImage(std::array<ilmType*, availableChannels>& pixelsArrayIlm, const asset::ICPUImage* image, const char* fileName)
+		{
+			auto getIlmType = [&]()
+			{
+				if (image->getCreationParameters().format == EF_R16G16B16A16_SFLOAT)
+					return PixelType::HALF;
+				else if (image->getCreationParameters().format == EF_R32G32B32A32_SFLOAT)
+					return PixelType::FLOAT;
+				else if (image->getCreationParameters().format == EF_R32G32B32A32_UINT)
+					return PixelType::UINT;
+			};
+
+			Header header(image->getCreationParameters().extent.width, image->getCreationParameters().extent.height);
+			const PixelType pixelType = getIlmType();
+			FrameBuffer frameBuffer;
+
+			const auto& width = image->getCreationParameters().extent.width;
+			const auto& height = image->getCreationParameters().extent.height;
+			const auto pitch = image->getRegions().begin()->bufferRowLength;
```

Comment:
this will be potentially different for every region, this is why you need to reserve the correct image sized buffer and copy pixels manually (construct the image from regions)

Reply:
there is only one region :D

Reply:
no goddamn it, I can give you an image with 50 regions in the spec and you need to be able to save it!

Reply:
to fix